Title for push notifications

wkhaksar opened the pull request:
With traccar APIs currently we are sending only "display messages" to Firebase, then it is automatically displayed in receiver's device while the app is in background. 

The conecpt [docs ](https://firebase.google.com/docs/cloud-messaging/concept-options) from Firebase says: FCM automatically displays the message to end-user devices on behalf of the client app.

So, with "display messages" it is good to fullfill the "title" property of the notification object as well, otherwise the notification message will be having only app-icon while displaying the message when the app is in background.

I am thinking, we can define a global title for the notification messages, if it exists then will be transferred along with the message body.

wkhaksar commented:
I added "**title**" support for short messages through templates.
Currently it is used with Firebase and Pushover as they are supporting titles, rest other channels are only using "**body**".

tananaev commented on src/main/java/org/traccar/notification/ShortMessage.java:
```diff
@@ -0,0 +1,36 @@
+ */
+package org.traccar.notification;
+
+public class ShortMessage {
```
Do we need this new model? Now that short and long messages have the same format, I think we should be able to simplify many things.

wkhaksar replied:
Fixed.

wkhaksar replied:
With a new commit, I removed it and renamed the "FullMessage" to a common name "Message" and used it for formatting both full and short messages. Also, with a few other renaming to make them distinguishable.

tananaev commented on src/main/java/org/traccar/notification/NotificationFormatter.java:
```diff
@@ -63,7 +63,7 @@ public static FullMessage formatFullMessage(long userId, Event event, Position p
     }
 
-    public static String formatShortMessage(long userId, Event event, Position position) {
+    public static ShortMessage formatShortMessage(long userId, Event event, Position position) {
```
Same here. We probably don't need a separate method.

wkhaksar replied:
This has been merged into "FormatMessage" in 9fd24af.

tananaev commented on templates/short/alarm.vm:
```diff
@@ -1 +1,2 @@
+#set($title = "Alarm")
```
Do all these titles match email titles?

wkhaksar replied:
In a new commit, copied subjects from "full" templates into "short" templates.

wkhaksar commented:
In this new commit 9fd24af :
1. Removed the new model "ShortMessage", so now we have only a common model **Message** instead of "FullMessage" and "ShortMessage". 
2. Merged "formatFullMessage" and "formatShortMessage" into a signle method: **formatMessage**.
3. Renamed "Message" models for Firebase and Pushover into **Payload**.
4. Copied subjects from "full templates" into "short templates".

tananaev commented on src/main/java/org/traccar/notificators/NotificatorTelegram.java:
```diff
@@ -74,15 +75,15 @@ public NotificatorTelegram() {
-            @Override
-            public void completed(Object o) {
-            }
+                    @Override
+                    public void completed(Object o) {
+                    }
```
Can you please revert these formatting changes.

tananaev commented on src/main/java/org/traccar/notificators/NotificatorPushover.java:
```diff
@@ -74,23 +77,26 @@ public void sendSync(long userId, Event event, Position position) {
             return;
         }
 
-        Message message = new Message();
```
I think it's better to rename `org.traccar.notification.Message` to something like `NotificationMessage`. That way we don't need to rename everything else.

tananaev commented on src/main/java/org/traccar/notification/TextTemplateFormatter.java:
```diff
@@ -71,16 +71,12 @@ public static Template getTemplate(String name, String path) {
-    }
-
-    private static String formatMessage(
+    private static String format(
```
Is this still needed as a separate method?

tananaev reviewed (commented):
I think we are almost there. Added minor comments about naming and formatting.

wkhaksar commented:
Please have a look from 2c4b4cd, your recommendations has been applied.

tananaev commented on src/main/java/org/traccar/notification/TextTemplateFormatter.java:
```diff
@@ -71,21 +71,11 @@ public static Template getTemplate(String name, String path) {
         StringWriter writer = new StringWriter();
         getTemplate(name, templatePath).merge(velocityContext, writer);
-        return writer.toString();
+        String formattedMessage = writer.toString();
+        return new NotificationMessage((String) velocityContext.get("subject"), formattedMessage);
```
```suggestion
        return new NotificationMessage((String) velocityContext.get("subject"), writer.toString());
```

tananaev commented on src/main/java/org/traccar/notificators/NotificatorFirebase.java:
```diff
@@ -77,15 +83,15 @@ public void sendSync(long userId, Event event, Position position) {
-                    LOGGER.warn("Firebase notification error", throwable);
-                }
-            });
+                        @Override
```
Looks like formatting is not reverted here.

wkhaksar replied:
Okat, I will try to merge it from master.

tananaev replied:
Why do you need to merge it? Just revert this change.

wkhaksar replied:
As it is automatically formatted by IDE so the overall file is affected.

wkhaksar replied:
Theser are reverted in 7d5a18e.

tananaev commented on src/main/java/org/traccar/notificators/NotificatorFirebase.java:
```diff
@@ -66,8 +69,11 @@ public void sendSync(long userId, Event event, Position position) {
+
             Notification notification = new Notification();
-            notification.body = NotificationFormatter.formatShortMessage(userId, event, position).trim();
+            notification.title= shortMessage.getSubject();
```
```suggestion
            notification.title = shortMessage.getSubject();
```

tananaev commented on src/main/java/org/traccar/notificators/NotificatorTelegram.java:
```diff
@@ -98,12 +99,14 @@ private LocationMessage createLocationMessage(String messageChatId, Position pos
     @Override
     public void sendSync(long userId, Event event, Position position) {
         User user = Context.getPermissionsManager().getUser(userId);
+        NotificationMessage shortMessage = NotificationFormatter.formatMessage(userId, event, position,"short");
```
```suggestion
        NotificationMessage shortMessage = NotificationFormatter.formatMessage(userId, event, position, "short");
```

tananaev commented on src/main/java/org/traccar/api/resource/PasswordResource.java:
```diff
@@ -53,8 +53,8 @@ public Response reset(@FormParam("email") String email) throws SQLException, Mes
                 velocityContext.put("token", token);
-                FullMessage message = TextTemplateFormatter.formatFullMessage(velocityContext, "passwordReset");
-                Context.getMailManager().sendMessage(userId, message.getSubject(), message.getBody());
+                NotificationMessage fullMessage = TextTemplateFormatter.formatMessage(velocityContext, "passwordReset","full");
```
```suggestion
                NotificationMessage fullMessage = TextTemplateFormatter.formatMessage(velocityContext, "passwordReset", "full");
```

wkhaksar commented:
So, formatting reverted and your suggested changes has been also accepted.

tananaev commented:
Looks like checks are failing. Make sure you run `gradle check` locally before pushing the changes.

wkhaksar commented:
Resolved `gradle check` errors locally in commit c370aa6.

tananaev commented:
Merged, thank you for the contribution.
